File: Problems/Algorithms/934. Shortest Bridge/shortest_bridge.py
from collections import deque


class Solution:
    def shortestBridge(self, grid: List[List[int]]) -> int:
        n, m = len(grid), len(grid[0])
        neighbors = [[0, 1], [0, -1], [1, 0], [-1, 0]]

        queue = deque()

        # transform one island using a dfs
        def dfs(i: int, j: int):
            if grid[i][j] == 1:
                queue.append((i, j, 0)) # 0 is the distance to the other island
                grid[i][j] = 2

                for neighbor in neighbors:
                    r, c = i + neighbor[0], j + neighbor[1]
                    if self.is_valid(n, m, r, c):
                        dfs(r, c)

        flag = False
        for i in range(n):
            for j in range(m):
                if grid[i][j] == 1:
                    dfs(i, j)
                    flag = True
                    break
            if flag: break

        # now find the distance using a bfs
        visited = set()
        while queue:
            i, j, dist = queue.popleft()

            if grid[i][j] == 1:
                return dist - 1
            else:
                for neighbor in neighbors:
                    r, c = i + neighbor[0], j + neighbor[1]
                    if self.is_valid(n, m, r, c):
                        if (r, c) not in visited:
                            visited.add((r, c))
                            queue.append((r, c, dist+1))

        return -1

    @staticmethod
    def is_valid(n: int, m: int, r: int, c: int) -> int:
        return r >= 0 and c >= 0 and r < n and c < m

File: Problems/Algorithms/934. Shortest Bridge/ShortestBridge.kt
class Solution {
    fun shortestBridge(grid: Array<IntArray>): Int {
        val n = grid.size
        val m = grid[0].size
        
        val neighbors = arrayOf(intArrayOf(0, 1), intArrayOf(0, -1), intArrayOf(1, 0), intArrayOf(-1, 0))
        
        val queue = ArrayDeque<IntArray>()
        var flag = false
        for (i in 0..n-1) {
            for (j in 0..m-1) {
                if (grid[i][j] == 1) {
                    dfs(queue, grid, neighbors, n, m, i, j)
                    flag = true
                    break
                }
            }
            if (flag) break
        }

        val visited = Array(n) { IntArray(m) { 0 } }
        return bfs(queue, grid, visited, neighbors, n, m)
    }
    
    private fun dfs(queue: ArrayDeque<IntArray>, grid: Array<IntArray>, neighbors: Array<IntArray>, n: Int, m: Int, r: Int, c: Int): Unit {
        if (grid[r][c] == 1) {
            grid[r][c] = 2
            queue.add(intArrayOf(r, c, 0))
            
            for (neighbor in neighbors) {
                val x = r + neighbor[0]
                val y = c + neighbor[1]
                
                if (isValid(n, m, x, y)) {
                    dfs(queue, grid, neighbors, n, m, x, y)
                }
            }
        }
    }
    
    private fun bfs(queue: ArrayDeque<IntArray>, grid: Array<IntArray>, visited: Array<IntArray>, neighbors: Array<IntArray>, n: Int, m: Int): Int {
        while (!queue.isEmpty()) {
            val curr = queue.removeFirst()
            val r = curr[0]
            val c = curr[1]
            val dist = curr[2]
            
            if (grid[r][c] == 1) {
                return dist - 1
            } else {
                for (neighbor in neighbors) {
                    val x = r + neighbor[0]
                    val y = c + neighbor[1]
                    
                    if (isValid(n, m, x, y) && visited[x][y] == 0) {
                        visited[x][y] = 1
                        queue.add(intArrayOf(x, y, dist+1))
                    }
                }
            }
        }
        
        return -1
    }
    
    private fun isValid(n: Int, m: Int, r: Int, c: Int): Boolean {
        return r >= 0 && r < n && c >= 0 && c < m
    }
}

